apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -509,98 +513,148 @@ private void ShutdownIfCompleted()
+            //   to make the batch go over its maximum allowed byte-count.
+            // The maximum size depends on whether local queueing is occuring.
+            // If _idleClients is empty, then all clients are currently sending, so local queueing is occuring;
+            // which means the current batch cannot be sent, even if it is full.
+            bool currentBatchIsFull = _idleClients.Count == 0 ?
+            _currentBatch.Messages.Count >= _batchMaxElementCount ||
+                (_currentBatch.Messages.Count > 0 && _currentBatch.ByteCount + extraByteCount >= _batchMaxByteCount) :
+            _currentBatch.Messages.Count >= _batchElementCountThreshold ||
+                (_currentBatch.Messages.Count > 0 && _currentBatch.ByteCount + extraByteCount >= _batchRequestByteThreshold);
+            if (currentBatchIsFull)
+            {
+                QueueCurrentBatch();
             }
         }
 
-        // Pre-condition: Must be locked
-        // Allow one message that is larger than requestByteThreshold
-        private bool CurrentBatchIsFull(bool isQueuing, int extraByteCount) => isQueuing ?
-                _currentBatch.Messages.Count >= _batchMaxElementCount || (_currentBatch.Messages.Count > 0 && _currentBatch.ByteCount >= _batchMaxByteCount) :
-                _currentBatch.Messages.Count >= _batchElementCountThreshold || (_currentBatch.Messages.Count > 0 && _currentBatch.ByteCount >= _batchRequestByteThreshold);
+        private void QueueCurrentBatch()
+        {
+            // Pre-condition: Must be locked
+            // Cancel the timeout for this batch.
+            _currentBatch.TimerCts.Cancel();
+            // Queue the batch ready for sending.
+            _batchesReady.Enqueue(_currentBatch);
+            // Mark that there is no current batch.
+            _currentBatch = null;
+            // Trigger send to server.
+            TriggerSend();
+        }
 
-        // Not "async void" to gain standard async exeption handling
-        private async Task TriggerSend()
+        private void TriggerSend()
         {
-            // Pre-condition: Must be locked - lock active up until first 'await'
+            // Pre-condition: Must be locked.
             // Start sending a batch if there's a batch to send, and a client to use to send it.
-            if (_batchesReady.Count > 0 && _idleClients.Count > 0)
+            if (_batchesReady.Count == 0 || _idleClients.Count == 0)
+            {
+                return;
+            }
+            // Remove client and batch from relevant queues.
+            var client = _idleClients.Dequeue();
+            var batch = _batchesReady.Dequeue();
+            _batchesInFlightCount += 1;
+            // Update flow-control counts.
+            _queueElementCount -= batch.Messages.Count;
+            _queueByteCount -= batch.ByteCount;
+            _taskHelper.Run(async () =>
```

Comment:
Extract this code into a separate method? (Maybe a local method, maybe not - but it's a significant chunk in its own right.)